// @cypress/chrome-recorder/dist/CypressStringifyExtension.js
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _CypressStringifyExtension_instances, _CypressStringifyExtension_appendStepType, _CypressStringifyExtension_appendChangeStep, _CypressStringifyExtension_appendClickStep, _CypressStringifyExtension_appendDoubleClickStep, _CypressStringifyExtension_appendHoverStep, _CypressStringifyExtension_appendKeyDownStep, _CypressStringifyExtension_appendNavigationStep, _CypressStringifyExtension_appendScrollStep, _CypressStringifyExtension_appendViewportStep;
import { StringifyExtension, StepType } from '@puppeteer/replay';
import { supportedRecorderKeys, } from './constants.js';
export class CypressStringifyExtension extends StringifyExtension {
    constructor() {
        super(...arguments);
        _CypressStringifyExtension_instances.add(this);
    }
    async beforeAllSteps(out, flow) {
        out.appendLine(`describe(${formatAsJSLiteral(flow.title)}, () => {`);
        out
            .appendLine(`it(${formatAsJSLiteral(`tests ${flow.title}`)}, () => {`)
            .startBlock();
    }
    async afterAllSteps(out) {
        out.appendLine('});').endBlock();
        out.appendLine('});');
    }
    async stringifyStep(out, step, flow) {
        __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendStepType).call(this, out, step, flow);
        if (step.assertedEvents) {
            // TODO: handle assertions
        }
    }
}
_CypressStringifyExtension_instances = new WeakSet(), _CypressStringifyExtension_appendStepType = function _CypressStringifyExtension_appendStepType(out, step, flow) {
    switch (step.type) {
        case StepType.Click:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendClickStep).call(this, out, step, flow);
        case StepType.DoubleClick:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendDoubleClickStep).call(this, out, step, flow);
        case StepType.Change:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendChangeStep).call(this, out, step, flow);
        case StepType.SetViewport:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendViewportStep).call(this, out, step);
        case StepType.Scroll:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendScrollStep).call(this, out, step, flow);
        case StepType.Navigate:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendNavigationStep).call(this, out, step);
        case StepType.KeyDown:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendKeyDownStep).call(this, out, step);
        case StepType.Hover:
            return __classPrivateFieldGet(this, _CypressStringifyExtension_instances, "m", _CypressStringifyExtension_appendHoverStep).call(this, out, step, flow);
        case StepType.KeyDown:
            return;
    }
}, _CypressStringifyExtension_appendChangeStep = function _CypressStringifyExtension_appendChangeStep(out, step, flow) {
    const cySelector = handleSelectors(step.selectors, flow);
    if (cySelector) {
        out.appendLine(`${cySelector}.type(${formatAsJSLiteral(step.value)});`);
    }
    out.appendLine('');
}, _CypressStringifyExtension_appendClickStep = function _CypressStringifyExtension_appendClickStep(out, step, flow) {
    const cySelector = handleSelectors(step.selectors, flow);
    const hasRightClick = step.button && step.button === 'secondary';
    if (cySelector) {
        hasRightClick
            ? out.appendLine(`${cySelector}.rightclick();`)
            : out.appendLine(`${cySelector}.click();`);
    }
    else {
        console.log(`Warning: The click on ${step.selectors[0]} was not able to be exported to Cypress. Please adjust your selectors and try again.`);
    }
    if (step.assertedEvents) {
        step.assertedEvents.forEach((event) => {
            if (event.type === 'navigation') {
                out.appendLine(`cy.location("href").should("eq", "${event.url}");`);
            }
        });
    }
    out.appendLine('');
}, _CypressStringifyExtension_appendDoubleClickStep = function _CypressStringifyExtension_appendDoubleClickStep(out, step, flow) {
    const cySelector = handleSelectors(step.selectors, flow);
    if (cySelector) {
        out.appendLine(`${cySelector}.dblclick();`);
    }
    else {
        console.log(`Warning: The click on ${step.selectors[0]} was not able to be exported to Cypress. Please adjust your selectors and try again.`);
    }
    out.appendLine('');
}, _CypressStringifyExtension_appendHoverStep = function _CypressStringifyExtension_appendHoverStep(out, step, flow) {
    const cySelector = handleSelectors(step.selectors, flow);
    if (cySelector) {
        out.appendLine(`${cySelector}.trigger("mouseover");`);
    }
    out.appendLine('');
}, _CypressStringifyExtension_appendKeyDownStep = function _CypressStringifyExtension_appendKeyDownStep(out, step) {
    const pressedKey = step.key.toLowerCase();
    if (pressedKey in supportedRecorderKeys) {
        const keyValue = supportedRecorderKeys[pressedKey];
        out.appendLine(`cy.type(${formatAsJSLiteral(`{${keyValue}}`)});`);
        out.appendLine('');
    }
}, _CypressStringifyExtension_appendNavigationStep = function _CypressStringifyExtension_appendNavigationStep(out, step) {
    out.appendLine(`cy.visit(${formatAsJSLiteral(step.url)});`);
    out.appendLine('');
}, _CypressStringifyExtension_appendScrollStep = function _CypressStringifyExtension_appendScrollStep(out, step, flow) {
    if ('selectors' in step) {
        out.appendLine(`${handleSelectors(step.selectors, flow)}.scrollTo(${step.x}, ${step.y});`);
    }
    else {
        out.appendLine(`cy.scrollTo(${step.x}, ${step.y});`);
    }
    out.appendLine('');
}, _CypressStringifyExtension_appendViewportStep = function _CypressStringifyExtension_appendViewportStep(out, step) {
    out.appendLine(`cy.viewport(${step.width}, ${step.height});`);
    out.appendLine('');
};
function formatAsJSLiteral(value) {
    return JSON.stringify(value);
}
function filterArrayByString(selectors, value) {
    return selectors.filter((selector) => value === 'aria/'
        ? !selector[0].includes(value)
        : selector[0].includes(value));
}
function handleSelectors(selectors, flow) {
    // Remove Aria selectors in favor of DOM selectors
    const nonAriaSelectors = filterArrayByString(selectors, 'aria/');
    let preferredSelector;
    // Give preference to user-specified selectors
    if (flow.selectorAttribute) {
        preferredSelector = filterArrayByString(nonAriaSelectors, flow.selectorAttribute);
    }
    if (preferredSelector && preferredSelector[0]) {
        return `cy.get(${formatAsJSLiteral(preferredSelector[0][0])})`;
    }
    else {
        return `cy.get(${formatAsJSLiteral(nonAriaSelectors[0][0])})`;
    }
}
function assertAllValidStepTypesAreHandled(step) {
    console.log(`Warning: Cypress does not currently handle migrating steps of type: ${step.type}. Please check the output to see how this might affect your test.`);
}
